### PR TITLE
Add Rename Symbol Capability

### PR DESCRIPTION
**Pull request by mcecode**
Hi, I'd like to give implementing  #161 a try, though it might take some time since I'm new to language servers and renaming in Bash doesn't seem to be as straightforward as in other languages.  I have `onPrepareRename` working already, so hopefully `onRenameRequest` won't be too difficult to implement.

For global defined variables, I'm planning to base the rename on the  `includeAllWorkspaceSymbols` config. If it's set to `false`, then the symbol would only be renamed within files linked by sourcing. If it's set to `true`, then the symbol would be renamed throughout the whole workspace.

For variables that are not defined, renaming would only happen within the file. I'm thinking this could be useful for renaming environment variables like if someone wants to change `$HOME` to `$PWD` for some reason.

~~For special variables, I'm thinking `$1` to `$9` could be renamed since it's common to assign these to variables with more descriptive names. It could be useful, for example, if someone created a quick script or function and now want to clean it up, they could just rename `$1` to `$myVar` and assign `myVar="$1"` at the top of the script or function. I don't think other special variables like `$@` or `$#` should be renamed since they already have meaning in and of themselves.~~

Feedback on the approach I'm thinking of taking and the work I've done so far would be much appreciated :)

### Todos:

- [x] Add rename handlers
- [x] Implement `onPrepareRename`
- [x] Implement `onRenameRequest` for undeclared symbols
- [x] Implement `onRenameRequest` for scoped symbols
  - [x] Local variables inside functions
  - [x] Scoped variables and functions inside subshells
- [x] Implement `onRenameRequest` for global symbols
  - [x] Within the file
  - [x] Within the workspace
- [x] Test
  - [x] Fix failing
  - [x] `onPrepareRename` handler
  - [x] `onRenameRequest` handler
- [x] Refactor
- [x] Update any docs that reference rename symbol functionality (e.g., `README.md`)

### CODE REVIEW

**Comment by skovhus** [comment]
Thanks for looking into this! And great that you are opening a draft PR, makes it easier to help you. 

I do believe we have all the building blocks available to implement this feature – also note that a lot of the required functionality is available through https://github.com/bash-lsp/bash-language-server/blob/main/server/src/util/declarations.ts (which is what is used by the analyzer to get declarations). Declarations contain both environment variables, variables and functions.

> For special variables, I'm thinking $1 to $9 could be renamed since it's common to assign these to variables with more descriptive names. It could be useful, for example, if someone created a quick script or function and now want to clean it up, they could just rename $1 to $myVar and assign myVar="$1" at the top of the script or function. I don't think other special variables like $@ or $# should be renamed since they already have meaning in and of themselves.

I suggest descoping that for the first version (unless this is easy to implement).

**Comment by mcecode** [comment]
Thanks for the quick response! I'm still getting familiar with the code base so thanks for the tip on using the functionality in `declarations.ts`. 

I'll take your advice and descope my ideas about special variables for now since I'm not sure how complex it is to implement.

**Comment by mcecode** [comment]
Hi @skovhus, it took a bit of exploration and experimentation, but I think I'm more or less done with renaming not defined and scoped variables and functions, and I think renaming global variables and functions isn't too far off. Before I continue though, I'd like to get your feedback on what I've done and the approach I took so I know I'm going in the right direction.

## Cases covered

- Renaming not defined variables

   [not-defined.webm](https://github.com/bash-lsp/bash-language-server/assets/65783406/198882e2-e122-4da8-b3fe-ab882ae294c4)

- Renaming function scoped local variables

   [scoped-function.webm](https://github.com/bash-lsp/bash-language-server/assets/65783406/779c25c9-0d62-49bf-86f7-5435ce1bd102)

- Renaming subshell scoped functions and variables

   [scoped-subshell.webm](https://github.com/bash-lsp/bash-language-server/assets/65783406/a4f27249-a122-4306-b468-3adc53b2f406)

- Some scope awareness

   [scope-awareness-function.webm](https://github.com/bash-lsp/bash-language-server/assets/65783406/eba3e786-68b5-4d65-945a-88aef9044cc2)

   [scope-awareness-subshell.webm](https://github.com/bash-lsp/bash-language-server/assets/65783406/c722b1e0-ef8d-4422-a457-3431ce5ec7b7)

- Differentiation between functions and variables with the same name

   [differentiate-variables-and-functions.webm](https://github.com/bash-lsp/bash-language-server/assets/65783406/7150e523-7270-48f6-8684-46d771a0e0ac)

Though I think function nesting, subshells, and such are rarely needed or used, I opted to support them so that in those cases where a user does use them, they won't get too janky of an experience.

## Limitations

1. Variable names not typed as `variable_name` by tree-sitter-bash can't and don't get renamed. Examples of this are variables inside arithmetic expansions and C-style for loops.

   arithmetic-expansion-whole

   In the arithmetic expansion above, the whole `n+n` is typed as a `word` with a `command_name` parent. Even if separated apart, each character would just be seen as separate `word`s:

   arithmetic-expansion-separated

   To support these cases, we'd need to parse inside these constructs ourselves which would be quite complex to do.

2. Scope awareness breaks down for complex nesting and scopes.

   In the example below, scope-wise, `$var` inside `3` should not be renamed when renaming `var` inside `1`.

   [limitation-complex-nesting.webm](https://github.com/bash-lsp/bash-language-server/assets/65783406/c8b55319-6e26-4e3a-bf52-10f18057c578)

   In this other example, `$var` inside `calleeFunc` is not renamed when renaming `var` inside `callerFunc` even if, scope-wise, they are the same variable.

   [limitation-complex-scopes.webm](https://github.com/bash-lsp/bash-language-server/assets/65783406/e9f43760-e820-4709-9e8f-95ca495c2bba)

   This limitation comes from the heuristics I used when finding symbol instances. Though I could try to cover these and other edge cases, I think the effort-to-benefit ratio wouldn't be worth it right now as not a lot of users would run into them.

3. Only the first `variable_assignment` is caught per `declaration_command`, so in the example below, only `a="a"` would be caught.

   ```bash
   local a="a" b="b" c="c"
   ```

   Truthfully, I just forgot that multiple `variable_assignment`s can be done in one `declaration_command` while implementing `Analyzer#findOriginalDeclaration`, that's why it doesn't handle it. I don't think multiple `variable_assignment`s in one `declaration_command` is used often, so this can be left as is, but I can add it if you think it would be useful for users.

## Approach taken

The initial approach I took is as follows:

1. Get the `word` at the given position using `Analyzer#wordAtPointFromTextPosition`.
2. Find the `word`'s declaration using `Analyzer#findDeclarationsMatchingWord`.
3. Find all of the `word`'s occurrences using `Analyzer#findOccurrences`.

I quickly learned that though this approach worked well enough for renaming not defined variables, it broke down when dealing with things like local declarations and scopes. So, I ended up slowly adding methods to `Analyzer` as I revised my approach, which became:

1. Get the symbol, either a function or a variable, at the given position using `Analyzer#symbolAtPointFromTextPosition`
2. Find the symbol's original definition within the scope it's in using `Analyzer#findOriginalDeclaration`.
3. Find the definition's scope using `Analyzer#findParentScope`.
4. Find all of the symbol's occurrences inside that scope using `Analyzer#findOccurrencesWithin`.

I added methods instead of changing what's already there since I didn't want to interfere with the current functionality and diagnostics provided by the language server and risk breaking the current experience for users. However, I do think what I added could be incorporated later on by other handlers. For example, I think `Analyzer#findOccurrencesWithin`'s implementation could replace `Analyzer#findOccurrences`' which would give `onDocumentHighlight` and `onReferences` more accurate results.

That said, I do think I'm adding quite a bit of code, and I do realize that the onus of reviewing all of this would fall onto you. So, I wanted to know if you're alright with how I'm going about things and with all of the cases I'm trying to cover. Do you think I did too much? Should I tone down what I'm trying to support? Did I miss something and maybe I'm just reimplementing things that are already available in the code base? How could I make this easier for you to review later on?

File: server/src/analyser.ts [hunk 1]
```diff
@@ -332,6 +457,179 @@ export default class Analyzer {
     return locations
   }
 
+  /**
+   * A more scope-aware version of findOccurrences.
+   */
+  public findOccurrencesWithin({
```
Comment (skovhus) [inline server/src/analyser.ts:463]: I didn't look to carefully here, but it might be we can reuse some functionality from `utils/declarations`.

File: server/src/analyser.ts [hunk 2]
```diff
@@ -275,6 +275,131 @@ export default class Analyzer {
...
+
+      if (originalDeclaration && !continueSearching) {
+        break
+      }
+
+      boundary = parent.startPosition.row
+      parent = this.findParentScopeNode(uri, parent.startPosition, parent.endPosition)
+    }
+
+    // TODO: Handle global definitions
```
Comment (skovhus) [inline server/src/analyser.ts:394]: Do you mean declarations across files? `getGlobalDeclarations` should help with this.

**Review by skovhus — COMMENTED** [review]
Amazing work here. 👏 👏 👏  I quickly looked over the code and came with some suggestions. But I'm really happy about the level of communication you do here and the videos.

I would love if you invested in writing some high level unit tests covering all the cases we handle and do not handle.

File: server/src/server.ts [hunk 3]
```diff
@@ -717,6 +724,117 @@ export default class BashServer {
...
+      this.throwResponseError('You cannot rename a built-in command.')
+    }
+
+    if (this.executables.isExecutableOnPATH(symbol.word)) {
+      this.throwResponseError('You cannot rename an executable.')
+    }
+
+    if (ReservedWords.isReservedWord(symbol.word)) {
+      this.throwResponseError('You cannot rename a reserved word.')
+    }
```
[the remark this answers is not in the document]
Comment (mcecode) [reply server/src/server.ts:750]: You mean if someone wanted to do something like this?

```bash
echo="Hello"

echo() {
  printf '%s' "$1"
}

echo $echo
```

I tend to veer away from naming functions and variables using builtin and executable names in my scripts because it might cause confusion, so I thought it would be good to put some sort of guardrail for that. Then again, I guess language servers should be more general and less opinionated about these things, and if someone has a good use case for it, then we should give them the tools to work the way they want.

Renaming builtins and executables already work without these checks using the same rules as any other function or variable. If no definition is found, rename all instances. If a definition is found, rename all instances within the declaration's scope. See example below:

[echo-to-print.webm](https://github.com/bash-lsp/bash-language-server/assets/65783406/edfd8f56-fdb8-4ba7-94d2-7c2361a7c2f5)

I'll remove the checks to allow for this behavior.

File: server/src/analyser.ts [hunk 4]
```diff
@@ -689,4 +1042,18 @@ export default class Analyzer {
 
     return tree.rootNode.descendantForPosition({ row: line, column })
   }
+
+  private nodeAtPoints(
```
[the remark this answers is not in the document]
Comment (mcecode) [reply server/src/analyser.ts:1046]: Ah, when it comes to `nodeAtPoints`, the main difference between it and `nodeAtPoint` is that it can return a node for a whole range. For example, given:

```bash
func() {
  echo "Hello"
}
```

Using `nodeAtPoint` at the start of `func` would return the `word` `func`. On the other hand, using `nodeAtPoints` giving it the start and end of `func` would give the whole `function_definition`.

One place I use this is in `findOccurrencesWithin` to get the scope node given a `Range`.

Is this functionality already available somewhere else and I ended up reimplementing it? Or is there a better way of getting the whole node given a range?

File: server/src/analyser.ts [hunk 5]
```diff
@@ -672,6 +1000,31 @@ export default class Analyzer {
     return allSourcedUris
   }
 
+  private findParentScopeNode(
+    uri: string,
+    start: Parser.Point,
+    end: Parser.Point,
+  ): Parser.SyntaxNode | null {
+    const node = this.nodeAtPoints(uri, start, end)
```
[the remark this answers is not in the document]
Comment (mcecode) [reply server/src/analyser.ts:1008]: Oh yeah, you're right. I totally missed that the node is already available 😅

I'll improve the implementation of this and work on a better name and adding some JSDoc. Thanks!

File: server/src/analyser.ts [hunk 6]
```diff
@@ -525,6 +823,36 @@ export default class Analyzer {
...
+    )
+
+    if (!node) {
+      return null
+    }
+
+    if (
+      node.type === 'variable_name' ||
+      (node.type === 'word' &&
+        node.parent &&
```
[the remark this answers is not in the document]
Comment (mcecode) [reply server/src/analyser.ts:842]: Given the logic, do you mean this?

```js
node.type === 'variable_name' ||
(node.type === 'word' &&
  ['function_definition', 'command_name'].includes(node.parent?.type))
```

Either way, I'll try to improve the readability of these types of checks. There are other places I can think of that can be improved as well.

File: server/src/analyser.ts [hunk 7]
```diff
@@ -525,6 +823,36 @@ export default class Analyzer {
     )
   }
 
+  public symbolAtPointFromTextPosition(
```
[the remark this answers is not in the document]
Comment (mcecode) [reply server/src/analyser.ts:826]: This method is like `wordAtPointFromTextPosition` above it, but returns only specific symbols (i.e., variables and functions), doesn't necessarily have to be a part of a declaration. I created this to make sure what gets captured from a given position is something that can actually be renamed rather than a flag or other `word`s that can't.

Since you brought it up, I'm actually thinking of changing the signature of this to return a `SymbolKind` to stick with LSP semantics rather than making my own with the `type` property. I think doing that would help make  what is being returned more explicit.

**Comment by mcecode** [comment]
Thanks for the review, I appreciate it. I'll update the implementation based on your suggestions.

Don't worry about the tests, I've kept track of the cases covered and not and plan to write them in test form when the implementation is a bit more locked in. I've been adding and removing methods here and there during my exploration and didn't want to add and remove tests at the same time.

I'll just ping you again when I'm done with the implementation + tests. Hopefully it won't take me so long this time 😅 Thanks again!

File: server/src/analyser.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (skovhus) [reply server/src/analyser.ts:1046]: I now see the type signature is different. All good then. :)

File: server/src/analyser.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (skovhus) [reply server/src/analyser.ts:842]: Yeah replacing
```tsx
      node.type === 'variable_name' ||
      (node.type === 'word' &&
        node.parent &&
                (node.parent.type === 'function_definition' ||
          node.parent.type === 'command_name'))
          )
```
          
with something like:   

```tsx
node.type === 'variable_name' ||
(node.type === 'word' &&
  ['function_definition', 'command_name'].includes(node.parent?.type))
```

**Comment by skovhus** [comment]
It seems like this is pretty close! 👏

**Comment by mcecode** [comment]
Hi! Yeah, it's pretty close now. Taking me a while since I've only been able to chip at it little by little and I had to think through a bit on how workspace-wide renames should work, but the implementation's pretty much done. I just have to fix the failing tests due to snapshots not expecting `onPrepareRename` and `onRenameRequest` handlers, add the tests for renaming, and probably refactor a bit since some of the methods I've added have become quite long and unwieldy.

I'll probably make an updated write-up of the cases covered, limitations, and approach taken when I'm done as some things have changed since the last one I commented. Heads up though, I tried to use as much of what's already in the code base, but I wasn't able to use what's in `util/declarations.ts`  because it returns declarations based on the latest definition. The implementation I've come up with uses the original declaration/definition to gauge the scope of where to search for words that match. I'd be glad to know if I missed something and I can use it somewhere.

**Comment by skovhus** [comment]
> he implementation I've come up with uses the original declaration/definition to gauge the scope of where to search for words that match. I'd be glad to know if I missed something and I can use it somewhere.

No worries.

**Comment by mcecode** [comment]
Hi @skovhus, I'm finally done. There are a lot of lines added, but most of those are from tests and snapshots. I'm not sure if it's overkill, but I wanted to document as much behavior as I reasonably could while taking into account some edge cases, and that's what I ended up with.

Now, onto the updates. I'll be skipping visuals this time since I think the tests and example cases in `renaming.sh` should cover that.

## Cases covered

- Renaming undeclared symbols
- Renaming locally scoped symbols, local variables in functions and declared variables and functions in subshells, with some scope awareness
- Renaming globally scoped symbols, both file-wide and workspace-wide, taking into account the `includeAllWorkspaceSymbols` flag
- Differentiates between variables and functions with the same name
- Handling requests with non-renamable symbols (`onPrepareRename`) and invalid symbol names (`onRenameRequest`)

## Limitations

- Not all variables are typed by  tree-sitter-bash as `variable_name`, so those variables won't be renamed
- Scope awareness breaks down for complex scopes and nesting
- Only takes into account subshells created with `(` and ` )`, so, for example, subshell scopes created with the pipe (`|`) operator aren't recognized
- Doesn't take into account sourcing location and scope, so, for example, sourcing inside a subshell will affect symbols outside of that subshell

You can find tests that expound and give more examples of these in the "Edge or not covered cases" under the "onRenameRequest" describe block and some parts of the "onPrepareRename" describe block. Honestly though, there are probably more limitations that I'm just not aware of since there are so many cases that can be created with Bash.

## Approach taken

There are some changes from the original, but the overall steps are the same.

1. Get the symbol, either a function or a variable, at the given position using `Analyzer#symbolAtPointFromTextPosition`.
2. Find the symbol's original declaration and scope based on its original definition using `Analyzer#findOriginalDeclaration`. The general heuristic used here is that the original declaration should be within the same or at a higher scope as the symbol found in step 1 while also being higher up in the file.
3. Based on the original declaration and scope found in step 2, find all of the symbol's occurrences inside the scope, the file, or the whole workspace using  `Analyzer#findOccurrencesWithin`, taking into account things like scoping and the `includeAllWorkspaceSymbols` flag.

Thank you so much for your patience on this. Feel free to tell me if there's anything else I need to do or if I missed anything and I'll try to address it as soon as I can.

**Comment by mcecode** [comment]
Hi @Shane-XB-Qian, thanks for testing it on your setup. Glad it works well.

Unfortunately, yes, that case is an expected case since `b` is not typed by tree-sitter-bash as `variable_name` but rather as `word`, so it falls on the first limitation that I listed. I've added a while read loop test case since that's a common use case and updated the first test's name in "Edge or not covered cases" to make the limitation being tested more explicit.

**Review by skovhus — APPROVED** [review]

**Comment by skovhus** [comment]
Sorry for the late review. Amazing work here @mcecode – let me know if you have other ideas for improving the LSP. I would love some help maintaining this project...
